File: android/src/androidTest/java/org/ligi/passandroid/ThePassListSwiping.java
package org.ligi.passandroid;

import org.ligi.passandroid.model.PassStore;
import org.ligi.passandroid.ui.PassListActivity;

import javax.inject.Inject;

public class ThePassListSwiping extends BaseIntegration<PassListActivity> {

    public static final String CUSTOM_PROBE = "custom";
    @Inject
    PassStore passStore;

    public ThePassListSwiping() {
        super(PassListActivity.class);
    }

    @Override
    public void setUp() throws Exception {
        super.setUp();

        final TestComponent testComponent = DaggerTestComponent.builder().build();

        testComponent.inject(this);

        App.setComponent(testComponent);
        getActivity();
    }

    /*

    TODO figure out why this is flaky on some devices :-(

    @MediumTest
    public void testDialogOpensWhenSwipeRight() {
        onView(withId(R.id.pass_recyclerview)).perform(click());

        onView(withId(R.id.pass_recyclerview)).perform(RecyclerViewActions.actionOnItemAtPosition(0,swipeRight()));

        onView(withText("description")).perform(swipeRight());
        RecyclerViewActions.actionOnItemAtPosition(0,swipeRight());

        onView(withText(R.string.move_to_new_topic)).check(matches(isDisplayed()));
    }

    @MediumTest
    public void testWeCanMoveToTrash() {
        onView(withId(R.id.pass_recyclerview)).perform(click());
        onView(withText("description")).perform(swipeRight());

        onView(withId(R.id.suggestion_button_trash)).perform(click());

        onView(withText(R.string.topic_trash)).check(matches(isDisplayed()));
        assertThat(passStore.getClassifier().getTopics()).containsExactly(getActivity().getString(R.string.topic_trash));
    }


    @MediumTest
    public void testWeCanMoveToArchive() {
        onView(withId(R.id.pass_recyclerview)).perform(click());
        onView(withText("description")).perform(swipeRight());

        onView(withId(R.id.suggestion_button_archive)).perform(click());

        onView(withText(R.string.topic_archive)).check(matches(isDisplayed()));
        assertThat(passStore.getClassifier().getTopics()).containsExactly(getActivity().getString(R.string.topic_archive));
    }


    @MediumTest
    public void testWeCanMoveToCustom() {
        onView(withId(R.id.pass_recyclerview)).perform(click());
        onView(withText("description")).perform(swipeLeft());

        onView(withId(R.id.new_topic_edit)).perform(typeText(CUSTOM_PROBE));

        onView(withText(android.R.string.ok)).perform(click());

        assertThat(passStore.getClassifier().getTopics()).containsExactly(CUSTOM_PROBE);
    }


    @MediumTest
    public void testDialogOpensWhenSwipeLeft() {
        onView(withId(R.id.pass_recyclerview)).perform(click());
        onView(withId(R.id.pass_recyclerview)).perform(RecyclerViewActions.actionOnItemAtPosition(0, longClick()));
        onView(withId(R.id.pass_recyclerview)).perform(RecyclerViewActions.actionOnItemAtPosition(0, swipeLeft()));

        onView(withText(R.string.move_to_new_topic)).check(matches(isDisplayed()));
    }



    public static ViewAction swipeRight() {
        return actionWithAssertions(new GeneralSwipeAction(Swipe.SLOW,
                translate(GeneralLocation.CENTER_LEFT, -0.1f, 0),
                GeneralLocation.CENTER_RIGHT, Press.FINGER));
    }

    static CoordinatesProvider translate(final CoordinatesProvider coords,
                                         final float dx, final float dy) {
        return new CoordinatesProvider() {
            @Override
            public float[] calculateCoordinates(View view) {
                float xy[] = coords.calculateCoordinates(view);
                xy[0] += dx * view.getWidth();
                xy[1] += dy * view.getHeight();
                return xy;
            }
        };
    }
        */
}

File: android/src/androidTest/java/org/ligi/passandroid/ThePassListSwiping.kt
package org.ligi.passandroid

import androidx.recyclerview.widget.ItemTouchHelper
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.replaceText
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.*
import org.assertj.core.api.Assertions.assertThat
import org.junit.Rule
import org.junit.Test
import org.ligi.passandroid.ui.PassListActivity
import org.ligi.passandroid.ui.PassListFragment
import org.ligi.trulesk.TruleskIntentRule

const val CUSTOM_PROBE = "FOO_PROBE"

class ThePassListSwiping {

    @get:Rule
    val rule = TruleskIntentRule(PassListActivity::class.java) {
        TestApp.populatePassStoreWithSinglePass()
    }

    @Test
    fun testWeCanMoveToTrash() {
        fakeSwipeLeft()

        onView(withText(R.string.topic_trash)).perform(click())

        assertThat(TestApp.passStore.classifier.getTopics()).containsExactly(rule.activity.getString(R.string.topic_trash))
    }


    @Test
    fun testWeCanMoveToArchive() {
        fakeSwipeLeft()

        onView(withText(R.string.topic_archive)).perform(click())

        assertThat(TestApp.passStore.classifier.getTopics()).containsExactly(rule.activity.getString(R.string.topic_archive))
    }

    @Test
    fun testWeCanMoveToCustom() {

        fakeSwipeLeft()

        onView(withId(R.id.new_topic_edit)).perform(replaceText(CUSTOM_PROBE))

        onView(withText(android.R.string.ok)).perform(click())

        assertThat(TestApp.passStore.classifier.getTopics()).containsExactly(CUSTOM_PROBE)
    }


    @Test
    fun testDialogOpensWhenSwipeLeft() {
        fakeSwipeLeft()
        onView(withText(R.string.move_to_new_topic)).check(matches(isDisplayed()))
    }

    @Test
    fun testDialogOpensWhenSwipeRight() {

        fakeSwipeRight()

        rule.screenShot("move_to_new_topic_dialog")

        onView(withText(R.string.move_to_new_topic)).check(matches(isDisplayed()))
    }

    /*
      we have to fake swiping as the espresso methods swipeLeft and swipeRight made
      these tests flaky - more info here:
      http://stackoverflow.com/questions/35397439/swipe-tests-flaky-on-recyclerview
     */
    private fun fakeSwipe(dir: Int) {
        rule.activity.runOnUiThread {
            val fragment = rule.activity.supportFragmentManager.fragments.firstOrNull { it is PassListFragment } as PassListFragment
            fragment.onSwiped(0, dir)
        }
    }

    private fun fakeSwipeRight() = fakeSwipe(ItemTouchHelper.RIGHT)
    private fun fakeSwipeLeft() = fakeSwipe(ItemTouchHelper.LEFT)

}
